### PR TITLE
FINERACT-2148: update instalments interest with zero after charge off with interest recalculation enabled

### PR DESCRIPTION
**Pull request by oleksii-novikov-onix**
## Description

1. Update installment interest to zero after charge-off when interest recalculation is enabled.
2. Reprocess transactions after undoing the charge-off.
3. Covered by end-to-end test cases for when interest recalculation is enabled.

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java [hunk 1]
```diff
@@ -3260,7 +3275,41 @@ public CommandProcessingResult chargeOff(JsonCommand command) {
         final List<Long> existingTransactionIds = loan.findExistingTransactionIds();
         final List<Long> existingReversedTransactionIds = loan.findExistingReversedTransactionIds();
 
-        LoanTransaction chargeOffTransaction = LoanTransaction.chargeOff(loan, transactionDate, txnExternalId);
+        if (LoanChargeOffBehaviour.ZERO_INTEREST.equals(loan.getLoanProductRelatedDetail().getChargeOffBehaviour())) {
+            final List<LoanRepaymentScheduleInstallment> repaymentScheduleInstallments = loan.getRepaymentScheduleInstallments();
+
+            if (loan.isInterestRecalculationEnabled()) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: I dont think we need this. We are about to reproduce loan transactions in the line 3319 already. I think charge-off behaviour can easily be moved into the `AdvancedPaymentScheduleTransactionProcessor` and only handle there the changes. This way when reverse-replay situations occurs, Fineract will handle properly the recalculation of Charge-off and updating the repayment schedule.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java [hunk 2]
```diff
@@ -3902,4 +3952,56 @@ public void closeAsMarkedForReschedule(final Loan loan, final JsonCommand comman
...
+    private void calculatePartialPeriodInterest(final Loan loan, final LocalDate chargeOffDate) {
+        loan.getRepaymentScheduleInstallments().stream()
+                .filter(installment -> !installment.getFromDate().isAfter(chargeOffDate) && installment.getDueDate().isAfter(chargeOffDate))
+                .forEach(installment -> {
+                    final BigDecimal totalInterest = installment.getInterestOutstanding(loan.getCurrency()).getAmount();
+                    final long totalDaysInPeriod = ChronoUnit.DAYS.between(installment.getFromDate(), installment.getDueDate());
+                    final long daysTillChargeOff = ChronoUnit.DAYS.between(installment.getFromDate(), chargeOffDate);
+
+                    final BigDecimal interestTillChargeOff = totalInterest
+                            .divide(BigDecimal.valueOf(totalDaysInPeriod), 10, RoundingMode.HALF_UP)
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: You might wanna use the MathContext from Moneyhelper to be unified! Hardcoding values are not recommended!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java [hunk 3]
```diff
@@ -3902,4 +3952,56 @@ public void closeAsMarkedForReschedule(final Loan loan, final JsonCommand comman
...
+                    installment.updatePrincipal(defaultToZeroIfNull(installment.getPrincipal()).add(interestRemoved));
+                    installment.updateInterestCharged(interestTillChargeOff);
+                });
+    }
+
+    private BigDecimal defaultToZeroIfNull(final BigDecimal possibleNullValue) {
+        return possibleNullValue != null ? possibleNullValue : BigDecimal.ZERO;
+    }
+
+    private void updateLastInstallmentPrincipal(final Loan loan,
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: This looks a little overcomplicated...we have the loan entity on hand, so fetching repayment schedule from database might not be the recommended way to handle this.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java [hunk 4]
```diff
@@ -3902,4 +3952,56 @@ public void closeAsMarkedForReschedule(final Loan loan, final JsonCommand comman
...
+                            .divide(BigDecimal.valueOf(totalDaysInPeriod), 10, RoundingMode.HALF_UP)
+                            .multiply(BigDecimal.valueOf(daysTillChargeOff));
+
+                    final BigDecimal interestRemoved = totalInterest.subtract(interestTillChargeOff);
+                    installment.updatePrincipal(defaultToZeroIfNull(installment.getPrincipal()).add(interestRemoved));
+                    installment.updateInterestCharged(interestTillChargeOff);
+                });
+    }
+
+    private BigDecimal defaultToZeroIfNull(final BigDecimal possibleNullValue) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: MathUtil provides such capability.

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 5]
```diff
@@ -1588,8 +1588,10 @@ public List<LoanTransaction> retrieveListOfTransactionsForReprocessing() {
     }
 
     private static Predicate<LoanTransaction> loanTransactionForReprocessingPredicate() {
-        return transaction -> transaction.isNotReversed() && (transaction.isChargeOff() || transaction.isReAge()
-                || transaction.isAccrualActivity() || transaction.isReAmortize() || !transaction.isNonMonetaryTransaction());
+        return transaction -> (transaction.isNotReversed() && (transaction.isChargeOff() || transaction.isReAge()
+                || transaction.isAccrualActivity() || transaction.isReAmortize() || !transaction.isNonMonetaryTransaction()))
+                || (transaction.isReversed() && transaction.isChargeOff() && LoanChargeOffBehaviour.ZERO_INTEREST
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: I dont think we need this! Would you mind to remove?

Comment (oleksii-novikov-onix) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: Yes, you are right. I rechecked it and roll it back

**Review by adamsaghy — APPROVED** [review]
LGTM

**Pull request merged**